edc-client/src/main/java/com/turn/edc/router/RequestRouter.java

```diff
@@ -148,24 +150,38 @@ public void update(List<CacheInstance> availableInstances) {
 	 * Get the storage connection
 	 * If the connection not found, create a new one
 	 * */
-	public StorageConnection getConnection(CacheInstance source) 
+	public StorageConnection getConnection(CacheInstance source)
 			throws TimeoutException, IOException {
 		StorageConnection connection = routingMap.get(source.hashCode());
 
-		if (connection == null) {
-			logger.info("Existing connection not found, creating new connection: {}", source);
-			try {
-				connection = connectorFactory.create(
-						source.getHostAndPort().getHostText(),
-						source.getHostAndPort().getPort(),
-						TIMEOUT
-				);
-			} catch (IOException ioe) {
-				logger.error("Unable to establish new connection to {}" + source);
-				throw ioe;
+		if (connection != null) {
+			if (connection.isConnected()) {
+				return connection;
 			}
-			routingMap.put(source.hashCode(), connection);
+
+			logger.info("Throttle this request while connect is still initializing: {}", source);
+			throw new IOException("Connect is still initializing: " + source);
+		}
+
+		logger.info("Existing connection not found, creating new connection: {}", source);
+		connection = connectorFactory.create(
+				source.getHostAndPort().getHostText(),
+				source.getHostAndPort().getPort(),
+				TIMEOUT
+		);
+		routingMap.put(source.hashCode(), connection);
```

Comment:
How about only putting this connection into the map if connect() succeeded? This way, we won't have to remove it again later if the connection didn't succeed.

Reply:
We used to do that way. But it caused problem if it took long to get connected, like cross dc connection, while several clients are trying to create same connections. Then we will seeing lots of duplicated connections at the same time.